Duplicate entities on discovery

ehendrix23 opened the pull request:
## Description:

Clean-up coding for detecting duplicate entities (devices) upon discovery.

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

ehendrix23 committed:
Enhancements for DirecTV media player

Following enhancements have been made:

1. Added debug logging
2. Added ability to change channel using select_source service of the remote platform.
3. State will now show paused if a recorded program is paused, for live TV playing will always be returned.
4. Added the following attributes:
    a. media_position: current position of the media (in seconds)
    b. media_position_updated_at: timestamp when media_position was updated.
   c. source: current source (channel).
   d. media_isbeingrecorded: if current media is being recorded or not.
   e. media_rating: TV/Movie rating of the media
   f. media_recorded: if current media is recorded or live TV
   g. media_starttime: Timestamp media was aired

Reordered properties to follow same order as how they are in __init__.py of remote platform.

ehendrix23 committed:
Fixed error and cleaned up few items

Fixed an issue when determining if a program is recorded or not.
Cleaned up some coding.

ehendrix23 committed:
Fix issue in checking if DTV device is already configured

If a DTV device was configured before, then discovery would add this device again seperately if the name specified in the configuration is different from the name on the DTV.

This issue is fixed now. Part of the fix also ensure to allow multiple "primary" devices on the network to be discovered.
Further also added debug logging to the setup_platform.

ehendrix23 committed:
Further improvements

Some additional improvements related to handling the DATA_DIRECTV in hass.data.

ehendrix23 committed:
Fixed flake8 issue

Fixed flake8 issue

ehendrix23 committed:
Added available property

Added available property

ehendrix23 committed:
Updated to use get_locations()

Replaced doing the request for getLocations with the get_locations() API from DirectPy instead.

ehendrix23 committed:
Fix for checking if device is available

Fix for checking if device is available and small update to debug log message.

ehendrix23 commented on homeassistant/components/media_player/directv.py:
```diff
@@ -64,29 +65,59 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-                                      DEFAULT_PORT, loc["clientAddr"]])
+
+            _LOGGER.debug("Known devices: %s", known_devices)
+            for loc in resp.get("locations") or []:
```
@MartinHjelmare , 
Using getLocations() instead to retrieve the locations. Going through the data received is  done to confirm that the device has not already been created as an entity in HASS.

MartinHjelmare replied:
That's better. But we should not have to parse the json response and know the correct strings to look for to use the library. The library should implement an object model where we access the attributes of the objects. The objects and attributes of the objects should represent the features of the device.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

I think it could be ok to merge this for now, but in my opinion the library should take more responsibility before we can take any future changes to this integration.

ehendrix23 committed:
Fixed lint issue

Fixed lint issue with unused variable by adding ingore for it as this is for a enumerate

pvizeli commented:
Implement unique_id and the core is aware that there are no multiple entities around.

MartinHjelmare commented:
@pvizeli the platform still has to take care of not adding entities that are non unique. The core doesn't take care of this even with unique_id.

ehendrix23 commented on homeassistant/components/media_player/directv.py:
```diff
@@ -104,28 +136,43 @@ def __init__(self, name, host, port, device):
         self._last_position = None
         self._is_recorded = None
         self._assumed_state = None
+        self._available = False
```
@MartinHjelmare, no specific reason. I can move it into another PR. My only thought here was cause it is a small change to just add it here. Based on my understanding otherwise I need to wait till this PR is merged before then submitting the next one right?

MartinHjelmare replied:
You can base one on the other and submit both, but put one as WIP, and then rebase that one when the other is merged.

Or wait, like you say, and submit the second PR when the first is merged.

ehendrix23 commented on homeassistant/components/media_player/directv.py:
```diff
@@ -64,29 +64,61 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                                  str.title(loc["locationName"]),
+                                  host, loc["clientAddr"])
 
         except requests.exceptions.RequestException:
```
What is in between try/except block is to be executed if there is no issue with the request. If there is an issue with the request then the except block needs to be executed.

ehendrix23 commented:
@pvizeli, that is probably then why when I noticed this issue 1 of my DVRs went in as a duplicate (but with different entity id) whereas the other one did not. One of my DVRs has its name as "Family Rm" but in the HASS configuration I state "Family Room". On discovery it then is added to HASS as another entity ID.
The one named "Basement" both in the HASS configuration and the name on the DVR does not come in as duplicate.

ehendrix23 committed:
Updated try/except and removed available

Updated tr/except having the except by the statement we're doing except on.
Removed available, will be a different PR.

MartinHjelmare commented on homeassistant/components/media_player/directv.py:
```diff
@@ -64,29 +64,57 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            # Comparing based on host (IP) and clientAddr.
+            device_unknown = True
+            for device in known_devices:
+                if host in device and loc["clientAddr"] in device:
```
Instead of storing a list of four items `[name, host, port, client_address]` in `known_devices`, just store a tuple of `(host, client_address)`. Tuples are hashable so then we can use a single for loop to check membership by creating the tuple to search for from `host` and `client_address`.

We should avoid nesting for loops when possible.

```py
new_device = (host, client_address)
if new_device in known_devices:
   ...
```

ehendrix23 committed:
Updated known_devices to be tupples in a set

Updated known_devices to be a tupple in a set, removing loop to determine if client was already added.

ehendrix23 committed:
Merge branch 'dev' into Discovery_Duplicate_Devices_Issue

MartinHjelmare commented:
Maybe update the description since the other PR was merged before this one.
